docs/references/configurations.md
---
sidebar_position: 5
slug: /configurations
---

# Configurations

If you need to change or customize the behavior of the Infinity runtime, you can use a configuration file. By default, Infinity uses the default configuration. To specify your own configuration file, you can use the `--config path/to/config.toml` option.

When Infinity starts up, it determines and loads each item in the configuration in a specific order, and merges these configurations from different sources into a single configuration used by Infinity.

## Configuration Loading Order and Priority

When Infinity starts up, configurations are loaded and determined in the following order:

1. Default values embedded in the code.

2. Values from the configuration file provided with --config PATH (if exists).

The loading methods listed above have increasing priority, so a higher-priority source will override values loaded from a lower-priority source. If the provided configuration file is not found, Infinity will display a warning at startup. The whole configuration is also displayed upon startup.

## Configuration file example


```toml
[general]
# version of Infinity Server, defaults to version in source code
version                  = "0.4.0"
# timezone
time_zone                = "utc-8"
# number of worker threads, defaults to concurrent threads supported by compiler implementation
# range : [1, 16384]
cpu_limit                = 8

# Network-related configuration
[network]
# IP address of infinity server
server_address           = "0.0.0.0"

# Range for ports : [1024, 65535]
# Port of postgres server
postgres_port            = 5432
# Port of HTTP server
http_port                = 23820
# Port of Python SDK server
client_port              = 23817
# the maximum number of connections, defaults to 256
# range : [1, 65536]
connection_pool_size     = 128

# Log related configurations
[log]
# filename of log files
log_filename             = "infinity.log"
# directory of log files
log_dir                  = "/var/infinity/log"
# controls if the log should be printed to standard output
# range : {true|false} 
log_to_stdout            = false
# The maximum size of a single log file, defaults to 1GB
# range : at least 1GB
log_file_max_size        = "10GB"
# The number of rotating log files, defaults to 8
# range : [1, 655636]
log_file_rotate_count    = 10
# The level of output log, defaults to info
# range : {trace/debug/info/warning/error/critical}
log_level               = "info"

# storage-related configuration
[storage]
# The working directory of the persistence manager.
# The persistence manager maps block and index files to remote stored objects and manage the local cache.
persistence_dir          = "/var/infinity/persistence"

# periodically activates garbage collection:
# use "number + unit of time" to specify intervals
# s means seconds, for example "60s", 60 seconds
# m means minutes, for example "60m", 60 minutes
# h means hours, for example "1h", 1 hour
# set to "0s" to disable corresponding periodic tasks

# interval between index optimization tasks
# range : ["10s", "720h"]
optimize_interval        = "10s"
# interval between cleanup tasks
# range : ["10s", "720h"]
cleanup_interval         = "60s"
# interval between compaction tasks
# range : ["10s", "720h"]
compact_interval         = "120s"
# The threshold for flushing a single in-memory index to disk,
# when the number of rows the index includes exceeds this threshold,
# the system performs a flush operation for that in-memory index.
# range : [8192, 8388608]
mem_index_capacity       = 1048576

# buffer manager-related configurations
[buffer]
# the size of memory that buffer manager holds, defaults to 8GB
buffer_manager_size      = "4GB"
# the number of LRU caches in buffer manager
# range : [1, 100]
lru_num                  = 7
# When the required memory exceeds the size of the free memory in the buffer manager,
# the buffer manager dumps some of the its in-use memory to a temporary storage path specified by temp_dir.
temp_dir                 = "/var/infinity/tmp"
# The global memory index threshold triggers a flush to disk.
# When the memory used by all existing in-memory indices in the system exceeds this threshold
# the system performs a flush operation for all in-memory indices.
memindex_memory_quota   = "1GB"

# WAL related configuration
[wal]
# directory of WAL files
wal_dir                       = "/var/infinity/wal"
# interval between periodically full checkpoints
# range : ["0s", "720h"]
full_checkpoint_interval      = "86400s"
# interval between periodically delta(incremental) checkpoints
# range : ["0s", "720h"]
delta_checkpoint_interval     = "60s"

# the size threshold of WAL files for triggering compaction
# after a WAL file grows bigger than this threshold, the system will perform a compaction to it
# range : ["1KB", "1024GB"]
wal_compact_threshold         = "1GB"

# controls how WAL manager flush its logs to the disk
# flush_at_once: write and flush log each commit
# only_write: write log, OS control when to flush the log, default
# flush_per_second: logs are written after each commit and flushed to disk per second.
wal_flush                     = "only_write"

[resource]
# The resource files used by Infinity (such as the dictionary files used by the analyzer)
resource_dir                  = "/var/infinity/resource"
```

src/unit_test/main/config.cpp
// Copyright(C) 2023 InfiniFlow, Inc. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <gtest/gtest.h>
import base_test;

import stl;
import config;
import infinity_exception;
import third_party;
import status;
import compilation_config;

using namespace infinity;
class ConfigTest : public BaseTest {};

//static size_t GetAvailableMem() {
//    size_t pages = sysconf(_SC_PHYS_PAGES);
//    size_t page_size = sysconf(_SC_PAGE_SIZE); // Byte
//    return pages * page_size;
//}

TEST_F(ConfigTest, test1) {
    using namespace infinity;
    SharedPtr<String> path = nullptr;
    Config config;
    auto status = config.Init(path, nullptr);
    ASSERT_TRUE(status.ok());

    EXPECT_EQ(config.Version(), "0.4.0");
    EXPECT_EQ(config.TimeZone(), "UTC");
    EXPECT_EQ(config.TimeZoneBias(), 8);
    EXPECT_EQ(config.CPULimit(), std::thread::hardware_concurrency());

    // Network
    EXPECT_EQ(config.ServerAddress(), "0.0.0.0");
    EXPECT_EQ(config.PostgresPort(), 5432);
    EXPECT_EQ(config.HTTPPort(), 23820u);
    EXPECT_EQ(config.ClientPort(), 23817u);

    // Log
    EXPECT_EQ(config.LogFileName(), "infinity.log");
    EXPECT_EQ(config.LogDir(), "/var/infinity/log");
    EXPECT_EQ(config.LogFilePath(), "/var/infinity/log/infinity.log");
    EXPECT_EQ(config.LogToStdout(), false);
    EXPECT_EQ(config.LogFileMaxSize(), 1024l * 1024l * 1024l);
    EXPECT_EQ(config.LogFileRotateCount(), 8l);
    EXPECT_EQ(config.GetLogLevel(), LogLevel::kInfo);

    EXPECT_EQ(config.DataDir(), "/var/infinity/data");
    EXPECT_EQ(config.WALDir(), "/var/infinity/wal");

    // buffer
    EXPECT_EQ(config.BufferManagerSize(), 8 * 1024l * 1024l * 1024l);
    EXPECT_EQ(config.LRUNum(), 7);
    EXPECT_EQ(config.TempDir(), "/var/infinity/tmp");
    EXPECT_EQ(config.MemIndexMemoryQuota(), 4 * 1024l * 1024l * 1024l);
}

TEST_F(ConfigTest, test2) {
    using namespace infinity;
    SharedPtr<String> path = MakeShared<String>(String(test_data_path()) + "/config/infinity_conf.toml");
    Config config;
    auto status = config.Init(path, nullptr);
    ASSERT_TRUE(status.ok());

    EXPECT_EQ(config.Version(), "0.4.0");
    EXPECT_EQ(config.TimeZone(), "UTC");
    EXPECT_EQ(config.TimeZoneBias(), -8);

    EXPECT_EQ(config.CPULimit(), 2u);

    EXPECT_EQ(config.ServerAddress(), "127.0.0.1");
    EXPECT_EQ(config.PostgresPort(), 25432);
    EXPECT_EQ(config.HTTPPort(), 24821);
    EXPECT_EQ(config.ClientPort(), 24817);

    EXPECT_EQ(config.LogFileName(), "infinity.log");
    EXPECT_EQ(config.LogDir(), "/var/infinity/log");
    EXPECT_EQ(config.LogFilePath(), "/var/infinity/log/infinity.log");
    EXPECT_EQ(config.LogToStdout(), false);
    EXPECT_EQ(config.LogFileMaxSize(), 2 * 1024l * 1024l * 1024l);
    EXPECT_EQ(config.LogFileRotateCount(), 3l);
    EXPECT_EQ(config.GetLogLevel(), LogLevel::kTrace);

    EXPECT_EQ(config.DataDir(), "/var/infinity/data");
    EXPECT_EQ(config.WALDir(), "/var/infinity/wal");

    // buffer
    EXPECT_EQ(config.BufferManagerSize(), 3 * 1024l * 1024l * 1024l);
    EXPECT_EQ(config.LRUNum(), 8);
    EXPECT_EQ(config.TempDir(), "/var/infinity/tmp");
    EXPECT_EQ(config.MemIndexMemoryQuota(), 2 * 1024l * 1024l * 1024l);
}

TEST_F(ConfigTest, TestWrongParamNames) {
    using namespace infinity;
    SharedPtr<String> path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_invalid_param.toml");
    Config config;
    auto status = config.Init(path, nullptr);
    ASSERT_FALSE(status.ok()); 
}

TEST_F(ConfigTest, TestConfInvalidValues) {
    using namespace infinity;
    SharedPtr<String> path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_invalid_version.toml");
    Config config_invalid_version;
    auto status = config_invalid_version.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kMismatchVersion);

    Config config_invalid_timezone;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_invalid_timezone.toml");
    status = config_invalid_timezone.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidTimezone);

    Config config_invalid_server_address;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_invalid_server_address.toml");
    status = config_invalid_server_address.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidIPAddr);

    Config config_invalid_bytesize;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_invalid_bytesize.toml");
    status = config_invalid_bytesize.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidByteSize);

    Config config_invalid_log_level;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_invalid_log_level.toml");
    status = config_invalid_log_level.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidLogLevel);

    Config config_invalid_timeinfo;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_invalid_timeinfo.toml");
    status = config_invalid_timeinfo.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidTimeInfo);
}

TEST_F(ConfigTest, TestOutofRangeValues) {

    using namespace infinity;
    SharedPtr<String> path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_out_of_bound_bytesize.toml");
    Config config_bytesize;
    auto status = config_bytesize.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidConfig);

    Config config_number;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_out_of_bound_number.toml");
    status = config_number.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidConfig);
    
    Config config_timeinfo;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_out_of_bound_timeinfo.toml");
    status = config_timeinfo.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidConfig);

    Config config_timezone;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_out_of_bound_timezone.toml");
    status = config_timezone.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidConfig);

    Config config_walflush;
    path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_out_of_bound_walflush.toml");
    status = config_walflush.Init(path, nullptr);
    EXPECT_EQ(status.code(), ErrorCode::kInvalidConfig);
   
}

TEST_F(ConfigTest, TestValidValues) {
    using namespace infinity;
    SharedPtr<String> path = MakeShared<String>(String(test_data_path()) + "/config/test_conf_valid_value.toml");
    Config config;
    auto status = config.Init(path, nullptr);
    ASSERT_TRUE(status.ok());

    EXPECT_EQ(config.Version(), "0.4.0");
    EXPECT_EQ(config.TimeZone(), "UTC");
    EXPECT_EQ(config.TimeZoneBias(), -8);
    EXPECT_EQ(config.CPULimit(), 2);

    // Network
    EXPECT_EQ(config.ServerAddress(), "0.0.0.0");
    EXPECT_EQ(config.PostgresPort(), 5432);
    EXPECT_EQ(config.HTTPPort(), 23820u);
    EXPECT_EQ(config.ClientPort(), 23817u);
    EXPECT_EQ(config.ConnectionPoolSize(), 128);

    // Log
    EXPECT_EQ(config.LogFileName(), "infinity.log");
    EXPECT_EQ(config.LogDir(), "/var/infinity/log");
    EXPECT_EQ(config.LogFilePath(), "/var/infinity/log/infinity.log");
    EXPECT_EQ(config.LogToStdout(), false);
    EXPECT_EQ(config.LogFileMaxSize(), 1024l * 1024l * 1024l);
    EXPECT_EQ(config.LogFileRotateCount(), 10l);
    EXPECT_EQ(config.GetLogLevel(), LogLevel::kInfo);

    //storage
    EXPECT_EQ(config.DataDir(), "/var/infinity/data");
    EXPECT_EQ(config.OptimizeIndexInterval(), 10);
    EXPECT_EQ(config.CleanupInterval(), 60);
    EXPECT_EQ(config.CompactInterval(), 120);
    EXPECT_EQ(config.MemIndexCapacity(), 1048576);

    // buffer
    EXPECT_EQ(config.BufferManagerSize(), 4 * 1024l * 1024l * 1024l);
    EXPECT_EQ(config.LRUNum(), 7);
    EXPECT_EQ(config.TempDir(), "/var/infinity/tmp");
    EXPECT_EQ(config.MemIndexMemoryQuota(), 1024l * 1024l * 1024l);

    //wal
    EXPECT_EQ(config.WALDir(), "/var/infinity/wal"); 
    EXPECT_EQ(config.FullCheckpointInterval(), 86400l);
    EXPECT_EQ(config.DeltaCheckpointInterval(), 60l);
    EXPECT_EQ(config.WALCompactThreshold(), 1024l * 1024l * 1024l);
    //resource
    EXPECT_EQ(config.ResourcePath(), "/var/infinity/resource");
    //persistence
    EXPECT_EQ(config.PersistenceDir(), "/var/infinity/persistence");
}